R/deps_installation_proposal.R

```diff
@@ -299,6 +302,59 @@ get_refs_from_desc <- function(d) {
   res[res_idx]
 }
 
+#' Replace Remotes in the `desc` that have been resolved to a GitHub tag or are
+#' in CRAN
+#'
+#' Replaces any existing Remotes entry with the resolved GitHub tag from the
+#' `new_refs`.
+#'
+#' It keeps all the existing Remotes that have not been resolved in `new_refs`.
+#'
+#' @param d (`desc`) DESCRIPTION object
+#' @param new_refs (`list`) remote references that have been resolved and are
+#' being updated in `Config/Needs/verdepcheck`
+#' @keywords internal
+desc_remotes_cleanup <- function(d, new_refs) {
```

Comment:
I still struggle to fully understand what it does and why it is there.

1. What it does?
You have the following:
```
  new_remotes <- c(
    # Keep remotes (if the DESCRIPTION file is correct, this should have no elements)
    d$get_remotes()[!(remotes_pkg %in% new_ref_pkg)],
    # Modified remotes
    new_ref_remote
  )
  ```
  
`d$get_remotes()[!(remotes_pkg %in% new_ref_pkg)]` is a vector of remotes not present in config/needs/verdepcheck
`new_ref_remote` is vector with config/needs/verdepcheck refs for packages existing in remotes field
So essentially it's (conditionally) replacing remotes entries with config/needs/verdepcheck entries. OK seems that I get what it does while writing this comment.
UPDATE: I just came up with example where you actually remove some entries so it's not only about replacing one with another. Then I guess we can modify for loop with `if (cond1) then NULL else if (cond2) x else y`

2. Implementation:
- your function changes the order of remotes entries - don't yet know if it's relevant
- you call `$get_remotes()` twice - also not a big deal
- I was probably struggling because of names of var - there is `new_refs`, `new_ref_remote`, `new_ref_remote` and none of them indicates new references for remote field
How about a simple for loop:
```
lapply(
  pkgdepends::parse_pkg_refs(d$get_remotes()),
  function(x) {
    if (<some condition>) then x else config_needs_verdepcheck[get_pkg_name(config_needs_verdepcheck) == x$package]
  }
)
```
- (connected to the above) Let's call `desc_cond_set_refs` first and then `desc_remotes_cleanup` using one argument only indicating desc object. This could potentially simplify naming issue as well as it fits to the role better - you want to replace desc remotes field with another field (that is being extracted from the same desc object and not another argument)

3. Is it really needed?
Please have a look at `get_refs_from_desc` executed at the very beginning of each installation proposal constructor - it is that function which is responsible for preparing correct list of references for further analysis. I would even ask if we need a Remotes at all - as we already / should have everything we need in the config/needs/verdepcheck. That field can potentially have downstream implications so I am inclined to request to always clear it.